src/Microsoft.SqlTools.ServiceLayer/Connection/ConnectionService.cs

```diff
@@ -228,11 +230,93 @@ public ISqlConnectionFactory ConnectionFactory
         public virtual bool TryFindConnection(string ownerUri, out ConnectionInfo connectionInfo) => this.OwnerToConnectionMap.TryGetValue(ownerUri, out connectionInfo);
 
         /// <summary>
-        /// Validates the given ConnectParams object. 
+        /// Refreshes the auth token of a given connection, if needed
         /// </summary>
-        /// <param name="connectionParams">The params to validate</param>
-        /// <returns>A ConnectionCompleteParams object upon validation error, 
-        /// null upon validation success</returns>
+        /// <param name="ownerUri">The URI of the connection</param>
+        /// <returns>true if a refresh is needed,
+        /// false if the uri cannot be found, refresh is not needed,
+        /// refresh token has already been requested, or if refresh fails </returns>
+
+        internal async Task<bool> TryRefreshAuthToken(string ownerUri)
+        {
+            ConnectionInfo connInfo;
+            if (this.TryFindConnection(ownerUri, out connInfo))
+            {
+                // If not an azure connection, no need to refresh token
+                if (connInfo.ConnectionDetails.AuthenticationType != "AzureMFA")
+                {
+                    return false;
+                }
+                else
+                {
+                    // Check if token is expired or about to expire
+                    if (connInfo.ConnectionDetails.ExpiresOn - DateTimeOffset.Now.ToUnixTimeSeconds() < maxTolerance)
+                    {
+                        // Check if the token is updating already, in which case there is no need to request a new one
+                        if (this.IsConnected[ownerUri] == TokenState.Updating) {

```

Comment:
Given that @alanrenmsft is currently working on enabling async message processing we're going to have to start being more careful about stuff like this. You should add a lock (or even better use a concurrent object like ConcurrentDictionary)

(I don't believe there's a concurrent set so despite that being preferred you can just use a dictionary and have the value be just true or something throwaway if you can't find a better solution...)